### PR TITLE
DDP-5215 move to paginated exports

### PR DESCRIPTION
**Pull request by yufengwng**
## Context

Refactored data exports to use paginated queries (i.e. using limit and offset) so we're not storing the entire set of participants for a study in memory while doing exports. The code that orchestrates that is consolidated in `DataExportCoordinator` so it can be shared between the nightly export job and the on-demand export job.

I tested the pagination locally by setting a reasonably small number for `elasticsearchBatchSize`.

## Checklist

- [x] I have labeled the type of changes involved using the `C-*` labels.
- [x] I have assessed potential risks and labeled using the `R-*` labels.
- [x] I have considered error handling and alerts, and added `L-*` labels as needed.
- [x] I have considered security and privacy, and added `I-*` labels as needed
- [x] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.

## FUD Score

_Overall, how are you feeling about these changes?_

- [x] :relaxed: All good, business as usual!
- [ ] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?

_How does one observe these changes in a deployed system? Note that **user visible** encompasses many personas--not just patients and study staff, but also ops duty, your fellow devs, compliance, etc._

- [x] They are user-visible in dev as a regular user journey and require no additional instructions.
- [ ] Getting dev into a state where this is user-visible requires some tech fiddling. I have documented these steps in the related ticket.
- [ ] Requires other features before it's human visible. I have documented the blocking issues in jira.
- [ ] I have no idea how to demo this. Please help me!

## Testing

- [ ] I have written automated positive tests
- [ ] I have written automated negative tests
- [x] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [x] These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

### CODE REVIEW

File: pepper-apis/src/main/java/org/broadinstitute/ddp/event/HousekeepingTaskReceiver.java [hunk 1]
```diff
@@ -63,6 +63,9 @@ public void receiveMessage(PubsubMessage message, AckReplyConsumer reply) {
             case CLEANUP_TEMP_USERS:
                 handleCleanupTempUsers(message, reply);
                 break;
+            case CSV_EXPORT:
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/event/HousekeepingTaskReceiver.java:66]: @zyme I went ahead and added support for on-demand csv exports in this PR.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/export/DataExportCoordinator.java [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+            int fetchedSize = batch.size();
+            if (fetchedSize == 0) {
+                break;
+            }
+
+            boolean runSuccess = withAPIsTxn(handle -> {
+                boolean batchSuccess = true;
+                List<Participant> participants = exporter.extractParticipantDataSetByIds(handle, studyDto, batch);
+                LOG.info("Extracted {} participants for study {}", participants.size(), studyDto.getGuid());
+                if (indices.contains(ElasticSearchIndexType.PARTICIPANTS_STRUCTURED)) {
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/export/DataExportCoordinator.java:128]: Previously, the export was "index centric", meaning we bulk query for all participant in study and export them to each index. Now, the export is more "participant centric", in that we paginate through batches of participants and export them to the needed indices.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/export/DataExportCoordinator.java [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+            long elapsed = Instant.now().toEpochMilli() - start;
+            LOG.info("Finished csv export for study {} in {}s", studyGuid, elapsed / 1000);
+            return true;
+        } catch (Exception e) {
+            LOG.error("Error while running csv export for study {}, continuing", studyGuid, e);
+            return false;
+        }
+    }
+
+    boolean exportStudyToGoogleBucket(StudyDto studyDto, DataExporter exporter, Bucket bucket,
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/export/DataExportCoordinator.java:181]: Most of the orchestration here for csv exports was moved over from `StudyDataExportJob`.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/housekeeping/schedule/OnDemandExportJob.java [hunk 4]
```diff
@@ -78,67 +85,60 @@ public void execute(JobExecutionContext ctx) throws JobExecutionException {
...
+            Bucket bucket = bucketClient.getBucket(bucketName);
+            if (bucket == null) {
+                LOG.error("Could not find google bucket {}, skipping job {}", bucketName, getKey());
+                return;
+            }
+
+            coordinator.includeCsv(bucket);
         }
+
+        coordinator.export(studyDto);
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/housekeeping/schedule/OnDemandExportJob.java:142]: To do different kinds of exports, we just setup the coordinator and run `export()`.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/export/DataExportCoordinator.java [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+
+    private String buildExportBlobFilename(StudyDto study) {
+        return String.format("%s/%s", study.getName(), DataExporter.makeExportCSVFilename(study.getGuid(), Instant.now()));
+    }
+
+    <R, X extends Exception> R withAPIsTxn(HandleCallback<R, X> callback) throws X {
+        return TransactionWrapper.withTxn(TransactionWrapper.DB.APIS, callback);
+    }
+
+    private class PaginatedParticipantIterator implements Iterator<Participant> {
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/export/DataExportCoordinator.java:232]: Given how the csv export is currently setup and how we're mocking data in `DataExporterTest`, I opted to keep csv export mostly the same and instead control the pagination from within the iterator.

**Review by ssettipalli — DISMISSED** [review]
looks good to me.
I thought at some point we might bee deprecating CSV exports because pretty much all the data is available in DSM.
In local testing how was memory usage compared to previous code base

**Comment by yufengwng** [comment]
@ssettipalli I don't know what's the best way to measure how much memory something is taking up. There's also the fact of garbage collection. But I tried running DEV Angio (with 1400+ participants) locally. In previous code base, I _think_ there is about 10-15 MB of memory active throughout the export that GC cycles did not clean up (again, not sure if it's real active memory or the GC is just being lazy). While in this PR, I observed barely 1MB overhead during the export. Of course, there are memory spikes throughout export, but GC seems to run and able to reclaim good amount of memory.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/JdbiUserStudyEnrollment.java [hunk 6]
```diff
@@ -33,6 +33,17 @@
     @CreateSqlObject
     ActivityInstanceDao getActivityInstanceDao();
 
+    @SqlQuery("select user_id"
+            + "  from user_study_enrollment"
+            + " where study_id = :studyId"
+            + "   and valid_to is null"
+            + " order by user_id"
+            + " limit :limit offset :offset")
```
Comment (MocanaAtBroad) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/JdbiUserStudyEnrollment.java:41]: Here again, it probably the enrollment start time (or if not indexed, perhaps better the user_study_enrollment_id) might be a better order key.

**Review by MocanaAtBroad — APPROVED** [review]

**Pull request merged**